py/dml/codegen.py

```diff
@@ -1244,87 +1239,168 @@ def check_varname(site, name):
 @statement_dispatcher
 def stmt_local(stmt, location, scope):
     # This doesn't occur in DML 1.0
-    [name, asttype, init] = stmt.args
-    if dml.globals.dml_version == (1, 2) and not dml.globals.compat_dml12:
-        check_varname(stmt.site, name)
-    (struct_decls, etype) = eval_type(
-        asttype, stmt.site, location, scope)
-    etype = etype.resolve()
-    rt = safe_realtype_shallow(etype)
-    if isinstance(rt, TArray) and not rt.size.constant and deep_const(rt):
-        raise EVLACONST(stmt.site)
-    init = get_initializer(stmt.site, etype, init, location, scope)
-
-    check_shadowing(scope, name, stmt.site)
-
-    sym = scope.add_variable(name, type = etype,
-                             site = stmt.site, init = init,
-                             static = False,
-                             stmt = True,
-                             make_unique=not dml.globals.debuggable)
-
-    return ([mkStructDefinition(site, t) for (site, t) in struct_decls]
-            + [sym_declaration(sym)])
+    [decls, inits] = stmt.args
+    method_call_init = False
+    if inits is None:
+        inits = [None] * len(decls)
+    elif (len(inits) == 1 and inits[0].kind == 'initializer_scalar'
+          and inits[0].args[0].kind == 'apply'):
+        method_call_init = True
+    elif len(decls) != len(inits):
+        raise ESYNTAX(stmt.site, None,
+                      'wrong number of initializers:\n'
+                      + f'{len(decls)} variables declared\n'
+                      + f'{len(inits)} initializers specified')
+    stmts = []
+
+    def convert_decl(decl_ast):
+        (name, asttype) = decl_ast.args
+        if dml.globals.dml_version == (1, 2) and not dml.globals.compat_dml12:
+            check_varname(stmt.site, name)
+        (struct_decls, etype) = eval_type(asttype, stmt.site, location, scope)
+        stmts.extend(mkStructDefinition(site, t) for (site, t) in struct_decls)
+        etype = etype.resolve()
+        rt = safe_realtype_shallow(etype)
+        if isinstance(rt, TArray) and not rt.size.constant and deep_const(rt):
+            raise EVLACONST(stmt.site)
+        check_shadowing(scope, name, stmt.site)
+        return (name, etype)
+
+    decls = list(map(convert_decl, decls))
+
+    def add_var(scope, name, typ, init = None,
+                make_unique=not dml.globals.debuggable):
+        sym = scope.add_variable(
+            name, type = typ, site = stmt.site, init = init, stmt = True,
+            make_unique=not dml.globals.debuggable)
+        stmts.append(sym_declaration(sym))
+        return sym
+
+    if method_call_init:
+        local_scope = Symtab(scope)
+        late_sym_adds = []
+        late_declares = []
+        tgts = []
+
+        for (name, typ) in decls:
+            if typ.const:
+                nonconst_typ = typ.clone()
+                nonconst_typ.const = False
+                sym = add_var(local_scope, '_tmp_' + name, nonconst_typ,
```

Comment:
why don't you just create a LocalSymbol like in the else clause?  This way you can remove local_scope. You can still enclose that in a function add_var if you like.